app/src/components/fineTunes/FineTuneContentTabs/General/InferenceCodeTabs/InferenceCodeTabs.tsx

```diff
@@ -0,0 +1,131 @@
+import { useMemo } from "react";
+
+import { type RouterOutputs } from "~/utils/api";
+import { useFineTune, useSelectedProject, useTrainingEntries } from "~/utils/hooks";
+import CopiableCodeTabs, { type CodeTab } from "./CopiableCodeTabs";
+
+const baseTabs: CodeTab[] = [
+  {
+    title: "cURL",
+    language: "bash",
+    code: `curl --request POST \\
+--url https://app.openpipe.ai/api/v1/chat/completions \\
+--header 'Authorization: Bearer {{TEMPLATED_OPENPIPE_API_KEY}}' \\
+--header 'Content-Type: application/json' \\
+--data '{
+"model": "openpipe:{{FINE_TUNE_MODEL_SLUG}}",
+"messages": {{TEMPLATED_MESSAGES}},
+"tool_choice": {{TEMPLATED_TOOL_CHOICE}},
+"tools": {{TEMPLATED_TOOLS}}
+}'`,
+  },
+  {
+    title: "Python",
+    language: "python",
+    code: `from openpipe import OpenAI
+
+client = OpenAI(
+  openpipe={"api_key": "{{TEMPLATED_OPENPIPE_API_KEY}}"}
+)
+
+completion = client.chat.completions.create(
+    model="openpipe:{{FINE_TUNE_MODEL_SLUG}}",
+    messages={{TEMPLATED_MESSAGES}},
+    tool_choice={{TEMPLATED_TOOL_CHOICE}},
+    tools={{TEMPLATED_TOOLS}},
+    openpipe={
+        "tags": {
+            "prompt_id": "counting",
+            "any_key": "any_value"
+        }
+    },
+)
+
+print(completion.choices[0].message)`,
+  },
+  {
+    title: "JavaScript",
+    language: "javascript",
+    code: `import OpenAI from "openpipe/openai";
+
+const client = new OpenAI({
+    openpipe: {
+        apiKey: "{{TEMPLATED_OPENPIPE_API_KEY}}",
+    }
+});
+
+const completion = await client.chat.completions.create({
+    model: "openpipe:{{FINE_TUNE_MODEL_SLUG}}",
+    messages: {{TEMPLATED_MESSAGES}},
+    tool_choice: {{TEMPLATED_TOOL_CHOICE}},
+    tools: {{TEMPLATED_TOOLS}},
+});
+
+console.log(completion?.choices[0]?.message);
+`,
+  },
+];
+
+const templateTabs = (
+  apiKey: string,
+  modelSlug: string,
+  entry?: RouterOutputs["datasetEntries"]["listTrainingEntries"]["entries"][number]["datasetEntry"],
+): CodeTab[] => {
+  let formattedMessages = "";
+
+  if (entry?.messages) {
+    const messages = entry.messages || [
+      {
+        role: "system",
+        content: "Count to 10",
+      },
+    ];
+    // add a tab to all but the first line
+    formattedMessages = JSON.stringify(messages, null, 4);
+  }
+
+  const formattedToolChoice = JSON.stringify(entry?.tool_choice ?? "auto", null, 4);
+
+  const formattedTools = JSON.stringify(entry?.tools ?? []);
+
+  return baseTabs.map((tab) => {
+    const templatedCode = tab.code;
+
+    const tabFormattedMessages =
+      tab.language === "bash"
+        ? formattedMessages.replace(/'/g, `'"'"'`)
```

Comment:
If we start running into issues where some edge cases aren't escaped properly might be a good idea to switch to a dedicated library for this like https://www.npmjs.com/package/shell-escape.

Reply:
I don't think we will, but if we do I'd prefer to use another package. This one would require us to write our own types, and isn't very ergonomic.